kmod/src/data.c

```diff
@@ -1583,30 +1587,40 @@ int scoutfs_data_fiemap(struct inode *inode, struct fiemap_extent_info *fieinfo,
 			break;
 		}
 
-		if (scoutfs_ext_can_merge(&cur, &ext)) {
+		if (scoutfs_ext_can_merge(&cur[pos], &ext)) {
 			/* merged extents could be greater than input len */
-			cur.len += ext.len;
+			cur[pos].len += ext.len;
 		} else {
-			ret = fill_extent(fieinfo, &cur, 0);
-			if (ret != 0)
-				goto unlock;
-			cur = ext;
+			cur[pos] = ext;
 		}
 
 		iblock = ext.start + ext.len;
+
+		/* max 16 items at a time */
+		if (++pos >= SCOUTFS_DATA_FIEMAP_MAX)
+			break;
 	}
 
-	if (cur.len)
-		ret = fill_extent(fieinfo, &cur, last_flags);
 unlock:
 	scoutfs_unlock(sb, lock, SCOUTFS_LOCK_READ);
 	up_read(&si->extent_sem);
 	inode_unlock(inode);
 
-out:
 	if (ret == 1)
 		ret = 0;
 
+	for (pos = 0; pos < SCOUTFS_DATA_FIEMAP_MAX; pos++) {
+		if (cur[pos].len) {
+			ret = fill_extent(fieinfo, &cur[pos], pos == (last_pos - 1) ? FIEMAP_EXTENT_LAST : 0);
```

Comment:
This loop rewrite has a bunch of issues.  It never merges 'cause it always advances pos after storing an extent in cur.  It can lose the _LAST flag if it hits enoent after emptying cur.  It loses the ret = 1 -> 0 translation.  I think it can loop for ever if iblock runs into last without setting last_pos or ret?  The different types of pos and last_pos is confusing.  At a high level, the fiemap interface is so annoying because we have to read the next extent before returning an extent so that we can find out if we have to set _LAST.  I think what'd be clearer is having a main outer loop like the current iblock <= last, then two inner loops.  One that merges and prepares extents with _next and the other which drains the prepared array by calling the kernel's fiemap_fill_next_extent().  It could use the kernel's fiemap_info struct for the allocated array storage so that the _next loop could set the LAST flag without needing the variable state (last_flags/last_pos) between the two loops.  But critically, the second loop would leave one extent remaining so that the next _next call can see enoent and set the _LAST flag on that remaining extent.

Reply:
I've completely rewritten this and the readdir one as well to be in line with these comments.